### PR TITLE
[MSFT] Add `MSFTModuleOp`

### PR DESCRIPTION
**Pull request by teqdruid**
Adds a moduleop to the MSFT dialect. For now, it's mostly like `hw.module` but adds `parameters` to represent the specific parameterization of a parameterized module. Also necessarily adds a `msft.output` since `hw.output` expects to be in a `hw.module`. 

Step 3 of #1755.

### CODE REVIEW

**Comment by teqdruid** [comment]
Large portions copied from `hw.module` and modified where appropriate. Removed the builders and verifiers. Will add them later.

File: include/circt/Dialect/MSFT/MSFTOps.td [hunk 1]
```diff
@@ -38,3 +38,83 @@ def InstanceOp : MSFTOp<"instance", [
       `:` functional-type($inputs, results)
   }];
 }
+
+def OneOrNoBlocksRegion : Region<
+  CPred<"::llvm::hasNItemsOrLess($_self, 1)">,
+  "region with at most 1 block">;
+
+def MSFTModuleOp : MSFTOp<"module",
```
Comment (cgyurgyik) [inline include/circt/Dialect/MSFT/MSFTOps.td:46]: I'm not totally familiar with this dialect, but if an InstanceOp can be either an instance of a MSFT Module or a HW Module, would it make sense to use [ParentOneOf](https://mlir.llvm.org/docs/Traits/#hasparent)?

**Comment by lattner** [comment]
Out of curiosity, why do you want a new moduleop?

**Comment by teqdruid** [comment]
@lattner See #1755 for some of the reasons. Essentially, we need a super set of the attributes (not reflected in this PR) and it would be convenient to have methods on the sugar c++ class to operate on those attributes. The only added attribute in this PR is the parameterization which this module is or was generated. (Each parameterized module will result in one of these per parameterization based on the instantiation parameters.)

It was getting cumbersome to rely on random attributes on `hw.module`. Also, there may be certain optimizations in the future (e.g. inlining) which we want to prevent or act on specially.

File: include/circt/Dialect/MSFT/MSFTOps.td (hunk 1 above)
Comment (teqdruid) [reply include/circt/Dialect/MSFT/MSFTOps.td:46]: You mean on the `InstanceOp`? Yeah, I think that makes sense, but not strictly related to this PR.

**Comment by mikeurbach** [comment]
I think this might make sense in the short term, but isn't @lattner working on supporting parameters on `hw.module` in a principled way? If so, is there still a need for `msft.module`?

**Review by mikeurbach — COMMENTED** [review]
The code looks fine to me.

**Review by mikeurbach — APPROVED** [review]

**Comment by mikeurbach** [comment]
We talked offline, I think this makes sense as a target for what PyCDE is doing in the short term. Once the `hw.module` work is complete, perhaps that can influence how PyCDE does parameterization.

**Comment by lattner** [comment]
> @lattner See #1755 for some of the reasons. Essentially, we need a super set of the attributes (not reflected in this PR) and it would be convenient to have methods on the sugar c++ class to operate on those attributes. The only added attribute in this PR is the parameterization which this module is or was generated. (Each parameterized module will result in one of these per parameterization based on the instantiation parameters.)

> It was getting cumbersome to rely on random attributes on hw.module. Also, there may be certain optimizations in the future (e.g. inlining) which we want to prevent or act on specially.

Cool, makes sense.

**Pull request merged**